core/src/main/java/org/infinispan/persistence/marshallers/EntryVersionMarshaller.java

```diff
@@ -0,0 +1,34 @@
+package org.infinispan.persistence.marshallers;
+
+import java.io.IOException;
+
+import org.infinispan.container.versioning.EntryVersion;
+import org.infinispan.container.versioning.NumericVersion;
+import org.infinispan.container.versioning.SimpleClusteredVersion;
+import org.infinispan.protostream.MessageMarshaller;
+
+public class EntryVersionMarshaller implements MessageMarshaller<EntryVersion> {
+   @Override
+   public EntryVersion readFrom(ProtoStreamReader reader) throws IOException {
+      EntryVersion version = reader.readObject("numeric", NumericVersion.class);
+      return version != null ? version : reader.readObject("clustered", SimpleClusteredVersion.class);
+   }
+
+   @Override
+   public void writeTo(ProtoStreamWriter writer, EntryVersion version) throws IOException {
+      if (version instanceof NumericVersion)
```

Comment:
Question: This seems overly complex. Is there no way to just write Object where you don't have to give it the type? For example what if this had 10 different implementations, would we have to do an instance of check and define a proto field for each? This seems very cumbersome.

Reply:
I don't think so @wburns as protobuf doesn't support inheritance. @anistor Any thoughts?